### DISCUSSION TITLE
Comparison linting between different types

### DISCUSSION THREAD

**Discussion by theFong** [microsoft/pyright#1411] [Q&A] [answered]
Hello!

Love using Pyright -- it has changed the way I write python.

I noticed that my Pyright doesn't warn/error when making comparisons between two different types. I have tested this in a clean repo with strict enabled.

```python
# pyright: strict
import typing

A = typing.NewType("A", str)
B = typing.NewType("B", int)


x = A("123")
y = B(234)

x == y  # no warn/error

class C:
    ...

class D:
    ...

C() == D()  # no warn/error
```

Is there an existing config to enable warning/error when making comparison between types? 

Thanks!

**Comment by erictraut** [comment, marked as answer]
Most classes support equality checks between arbitrary types. For example, the following code runs without an exception:
```python
x = 1
y = "hi"
print(x == y) # False
print(y == None) # False
```

This is not only legal at runtime, it's also specifically allowed by the `__eq__` and `__ne__` methods inherited from `object` by most classes.
```python
class object:
    ...
    def __eq__(self, o: object) -> bool: ...
    def __ne__(self, o: object) -> bool: ...
```

You're free to declare `__eq__` and `__ne__` methods on your custom class that accept narrower types than `object`. For example:
```python
class A:
    # Allow comparison only with other "A" instances.
    def __eq__(self, o: "A") -> bool: ...
    def __ne__(self, o: "A") -> bool: ...
```

**Reply by theFong** [reply]
This makes sense thanks 👍
